### ISSUE TITLE
automount is giving “Too many levels of symbolic links"

### ISSUE THREAD

**Issue by sampie** [systemd/systemd#17545]
Hi, 

I get “Too many levels of symbolic links" error when trying to access the folder, when I have following mount directive at fstab:

sami@lelu.local:/home/sami /mnt/node fuse.sshfs noauto,x-systemd.automount,_netdev,users,idmap=user,IdentityFile=/home/sami/.ssh/id_rsa,allow_other,reconnect 0  0

However, running manually "mount /mnt/node" works without problems. I have tried asking help from forums in case it is a misconfiguration instead of a bug, but got no responses:
https://askubuntu.com/questions/1286375/too-many-levels-of-symbolic-links-error-when-using-systemd

**Comment by raven-au** [comment]
Do you get anything related to this in the journal?

The most common cause of this is the call back to user space returns success but the mount hasn't been done for some reason.

**Comment by sampie** [comment]
When running "journalctl -u mnt-node.automount", I don't see any errors, but single "ls /mnt/node" seems to generate lots of entries:

marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 4955 (ls)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 1367 (gsd-housekeepin)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 1351 (gvfsd-trash)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 1351 (gvfsd-trash)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 4955 (ls)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 1351 (gvfsd-trash)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 1351 (gvfsd-trash)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 1351 (gvfsd-trash)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 1351 (gvfsd-trash)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 4955 (ls)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 1351 (gvfsd-trash)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 1351 (gvfsd-trash)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 4955 (ls)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 1351 (gvfsd-trash)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 1351 (gvfsd-trash)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 4955 (ls)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 1351 (gvfsd-trash)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 4955 (ls)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 4955 (ls)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 4955 (ls)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 1351 (gvfsd-trash)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 1351 (gvfsd-trash)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 4955 (ls)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 1351 (gvfsd-trash)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 1351 (gvfsd-trash)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 1351 (gvfsd-trash)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 1351 (gvfsd-trash)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 1351 (gvfsd-trash)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 1351 (gvfsd-trash)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 1351 (gvfsd-trash)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 4955 (ls)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 1351 (gvfsd-trash)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 4955 (ls)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 4955 (ls)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 4955 (ls)
marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 1351 (gvfsd-trash)

**Comment by poettering** [comment]
[Canned reply follows] When submitting a GitHub issue, please fill in the submission form you find in the issue creation page, i.e. provide the requested details, such as distribution and systemd versions. We ask you these details to keep frustration at a minimum, both for us, and for you. If you don't provide this information it makes things needlessly difficult to help you.

Also, please keep in mind: this is a bug and feature request tracker only. If you have a support question, i.e. wonder how to use a feature, or can't figure out how to do something best, please use a different forum, such as the systemd mailing list. Also, do not file bug reports unless you can reproduce the issue with one of the two most recently released *upstream* version of systemd. For everything older, please contact your downstream distribution instead.

Thank you for understanding!

**Comment by sampie** [comment]
Basically this issue can be reproduced by installing latest LTS Ubuntu 20.04 (systemd 245.4-4ubuntu3.3, Linux kernel 5.4.0-52-generic). +  fuse-package + sshfs and then adding a mount entry to fstab. 

I have reproduced this issue also with the latest Debian and Ubuntu 20.10.

**Comment by raven-au** [comment]
> Basically this issue can be reproduced by installing latest LTS Ubuntu 20.04 (systemd 245.4-4ubuntu3.3, Linux kernel 5.4.0-52-generic, x64 architecture). + fuse-package + sshfs and then adding a mount entry to fstab.

As @poettering says, keep in mind this is an upstream issues tracker.

You can't assume people looking at this will be familiar enough with Ubuntu package versions or that they will be willing to install them and survey the source of the packages to understand what they are looking at. And there's no fuse package version either not to mention possible changes made to it as well.

For example, 245.4-4ubuntu3.3 means nothing to me, the Ubuntu package maintainers could have added fix patches that may or may not be related to this so looking at the current systemd upstream source could be a wild goose chase.

Also, and I get that you can't be expected to know this, but you might not have actually reproduced it, I don't know from what you have described.

Probably the closest way to reproduce this would be to perform that sshfs mount on top of some other mount on your system, not on a directory that has no mount on it.

Even that doesn't fully reproduce it since the location this gets mounted on in systemd is an autofs file system mounted at that point in the file system but you can't realistically do that yourself.

Another problem is, by what you have said you imply this has previously functioned ok, but haven't given any information about package versions for which it did function ok. Even the Ubuntu systemd and fuse package versions would be something to go on if it has in fact worked ok previously.

Having said all that, historically, the upstream autofs package had problems with parsing sshfs mounts but that implies this hasn't worked previously.

And then there's the problem that a mount succeeded status has been returned when in fact the mount hasn't been done (which is usually, but not always, what leads to this message). That could be util-linux mount(8) not behaving or systemd not returning the correct status.

So please provide missing versions, including util-linux, and whether this has worked previously.

**Comment by raven-au** [comment]
> When running "journalctl -u mnt-node.automount", I don't see any errors, but single "ls /mnt/node" seems to generate lots of entries:
> 
> marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 4955 (ls)
> marras 07 18:50:02 ariana systemd[1]: mnt-node.automount: Got automount request for /mnt/node, triggered by 1367 (gsd-housekeepin)

Unfortunately this isn't quite what I was hoping to get.
Possibly the filter has resulted in excluding some messages I would have liked to see, but I'm not sure.

But it does give me a place to start I think, I'll have a quick look around and see if anything stands out.

**Comment by sampie** [comment]
Hi,

> You can't assume people looking at this will be familiar enough with Ubuntu package versions or 
> that they will be willing to install them and survey the source of the packages to understand what 
> they are looking at. And there's no fuse package version either not to mention possible changes 
> made to it as well.
> For example, 245.4-4ubuntu3.3 means nothing to me, the Ubuntu package maintainers could 
> have added fix patches that may or may not be related to this so looking at the current systemd 
> upstream source could be a wild goose chase.

I have described the systemd like I see it, like it is in my system. I don't know how to provide better version information. I don't know what fixes / patches Ubuntu developers might have added. I am a user. However, I can print the version with systemctl, but it prints also 245. I too, don't know what this means (and I am not familiar with systemd version numbering either). 
$ systemctl --version
systemd 245 (245.4-4ubuntu3.3)
+PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=hybrid

The fuse version reported by my system is 2.9.9-3.

> Also, and I get that you can't be expected to know this, but you might not have actually reproduced 
> it, I don't know from what you have described.
> Probably the closest way to reproduce this would be to perform that sshfs mount on top of some other 
> mount on your system, not on a directory that has no mount on it.
> Even that doesn't fully reproduce it since the location this gets mounted on in systemd is an 
> autofs file system mounted at that point in the file system but you can't realistically do that 
> yourself.

What I intended to say is that, as far as I see, all the systems I tested: Ubuntu 20.04, 20.10 and the latest Debian behaves the same, all giving “Too many levels of symbolic links" error with the same configuration. 

I am not sure, if I understood the technical explanation. A single "ls" is enough to reproduce this issue and prior to the ls, /mnt/node appears to be unmounted. However, I am wondering why journalctl shows so many " Got automount request " for a single ls. 

> Another problem is, by what you have said you imply this has previously functioned ok, but 
> haven't given any information about package versions for which it did function ok. Even 
> the Ubuntu systemd and fuse package versions would be something to go on if it has in 
> fact worked ok previously.

No. I have not ever got this to work. Is my fstab line correct? util-linux is showing version 2.34-0.1ubuntu9.1.

**Comment by raven-au** [comment]
> 
> No. I have not ever got this to work. Is my fstab line correct? util-linux is showing version 2.34-0.1ubuntu9.1.

Right, so I installed ubuntu 20.04.1 LTS in a VM.
All the package versions I have are the same.

I initially got the mount failure you saw that lead to that message.
I played around and got it to work but after reverting all the changes I did and making my fstab entry (essentially) the same as yours (except for mount point, public key path and host of course) I couldn't make it fail.

So I'm stumped.

One thing I'm pretty sure of is the mount command is failing when systemd tries to mount it.

What does "mount|grep <your mount point>" return, after a reboot and before you try and access the mount?

**Comment by raven-au** [comment]
> I am not sure, if I understood the technical explanation. A single "ls" is enough to reproduce this issue and prior to the ls, /mnt/node appears to be unmounted. However, I am wondering why journalctl shows so many " Got automount request " for a single ls.

Well, when the sshfs mount is not mounted, you should see something like this:
systemd-1 on /mnt/sshfs/centos7 type autofs (rw,relatime,fd=81,pgrp=1,timeout=0,minproto=5,maxproto=5,direct,pipe_ino=25547)

That's the autofs direct mount trigger that systemd uses to get the notification from the kernel to mount it when it's accessed.
The systemd automount unit must have been started in order for this mount to have been done by systemd.

So when this is initially setup I do a:
sudo systemctl daemon-reload
sudo systemctl restart mnt-sshfs-centos7.automount
to get it started in my case (and when I change fstab too).

That message, just before all those symlink messages, is an indication that the above trigger mount has been done otherwise systemd won't get a callback to perform the mount and that's the source of all those messages.

There's no mystery to the messages.

What happens is, on access, the kernel sends a notification asking systemd to perform the mount, if the mount command fails but a success is returned the kernel VFS continues trying to follow the mount until it reaches the kernel maximum number of symlinks that can be followed, that count being incremented each time the kernel tries to follow the mount. I guess it's a bit puzzling you get only one initial message, not sure about that, might be the way systemd handles the mounting.

That's why I say there's an incorrect return from systemd because an error return to the kernel will stop that loop from continuing.
But I'm not able to work out how the systemd mounting process works just jet so I can't work out how to fix that aspect of the problem.

So, in short, I'm pretty sure systemd is not using a correct status when it notifies the kernel upon mount invocation failure, and I think that needs to be fixed.

But that won't help you to work out why the mount invocation is failing which is what's needed for you to make it work.

**Comment by raven-au** [comment]
Btw,

If you really do want to use the systemd automount facility I recommend getting rid of that "users" option.
You really don't want to mount this thing manually or umount it manually either for that matter, that's likely to cause problems.

And a x-systemd.idle-timeout=<seconds> option would be sensible otherwise there's no real point in using an automount, you might as well just create a mount unit for the sshfs mount only.
That then makes the "reconnect" option questionable too.

And also you probably want a user="the id of the user accessing the mount" and a uid="the gid of the user accessing the mount" because this thing will be mounted by root and that will likely cause problems with user permissions. Not sure about the usefulness of that idmap=user option either ... it might ne needed for the uid.gid option, not sure myself.

**Comment by sampie** [comment]
Hi,

Thank you for taking the time to take a look at this problem!

Immediately after reboot:
$ mount |grep mnt
systemd-1 on /mnt/node type autofs (rw,relatime,fd=44,pgrp=1,timeout=0,minproto=5,maxproto=5,direct,pipe_ino=26803)

$ ls /mnt/node
ls: cannot access '/mnt/node': Too many levels of symbolic links

After ls, the mnt looks the same:
$ mount |grep mnt
systemd-1 on /mnt/node type autofs (rw,relatime,fd=44,pgrp=1,timeout=0,minproto=5,maxproto=5,direct,pipe_ino=26803)

(I do have sshfs package installed.)

I tried to remove the "user" option and rebooted, but I still get the error message. Later, I tried remove also "idmap" and "reconnect". Still, the error persists. If there is a good example entry for fstab, I would be interested to try that.

Earlier, I have tried to create a service file instead of fstab entry, but I could not make it work either. If there is a good example of a unit file for sshfs automount (that can be triggered by a normal user by accessing the folder), I would be interested to try that also.

**Comment by raven-au** [comment]
> 
> Immediately after reboot:
> $ mount |grep mnt
> systemd-1 on /mnt/node type autofs (rw,relatime,fd=44,pgrp=1,timeout=0,minproto=5,maxproto=5,direct,pipe_ino=26803)

Yep, that's how it's supposed to look.

> 
> $ ls /mnt/node
> ls: cannot access '/mnt/node': Too many levels of symbolic links

Yes, that's the ELOOP error returned from the kernel.
I really can't think of any other way that can happen except for the mount command systemd calls failing for some reason.

> 
> After ls, the mnt looks the same:
> $ mount |grep mnt
> systemd-1 on /mnt/node type autofs (rw,relatime,fd=44,pgrp=1,timeout=0,minproto=5,maxproto=5,direct,pipe_ino=26803)

Right, that's what you should see if the mount of the real file system fails.

> 
> (I do have sshfs package installed.)

Right, that's obviously needed.

Funny thing is that mine started to work right after I created a symlink to sshfs from mount.fuse.sshfs but continued to work after I removed it, no idea why.

I thought the mount.fuse.sshfs would be needed by the mount(8) -t fuse.sshfs but apparently not.
That's likely due to changes/improvements in fuse and the kernel.

I had also uncommented the user_allow_other in /etc/fuse.conf at that point but again once it worked adding the comment back didn't stop it from working, as you would expect since the systemd mount on demand is done with root privileges. 

It would be worth uncommenting the user_allow_other and using the sshfs command to mount this as a user.
You would need to get rid of the non-sshfs mount options and stop the systemd automount unit before doing this, for me that was "sudo systemctl stop mnt-sshfs-centos7.automount".

And you could always try creating that symlink too.

> 
> I tried to remove the "user" option and rebooted, but I still get the error message. Later, I tried remove also "idmap" and "reconnect". Still, the error persists. If there is a good example entry for fstab, I would be interested to try that.

Yea, that's only going to be a problem if it's after the first sshfs mount option otherwise it won't make any difference.

> 
> Earlier, I have tried to create a service file instead of fstab entry, but I could not make it work either. If there is a good example of a unit file for sshfs automount (that can be triggered by a normal user by accessing the folder), I would be interested to try that also.

For "mount on demand" you do need to use the systemd automount function.

**Comment by sampie** [comment]
Hi,

What I now tried was:
1) Commented out "user_allow_other" from /etc/fuse
2) Simplified fstab. It is now: "sami@lelu.local:/home/sami    /mnt/node fuse.sshfs   x-systemd.automount,IdentityFile=/home/sami.ssh/id_rsa          0  0"
3) Made a link:
/usr/bin/mount.fuse.sshfs -> /usr/bin/sshfs
4) Reloaded/restarted the unit:
sudo systemctl daemon-reload
sudo systemctl stop mnt-node.automount
sudo systemctl start mnt-node.automount

ls /mnt/node still gives the same error.

**Comment by raven-au** [comment]
> 
> ls /mnt/node still gives the same error.

What happens when you run:
sshfs -o idmap=user,identityfile=/home/sami.ssh/id_rsa sami@lelu.local:/home/sami  /mnt/node < /dev/null

**Comment by raven-au** [comment]
>     3. Made a link:
>        /usr/bin/mount.fuse.sshfs -> /usr/bin/sshfs

File system mount helpers live in /usr/sbin not /usr/bin.

**Comment by raven-au** [comment]
> > Is my fstab line correct?

snip ...

> I played around and got it to work but after reverting all the changes I did and making my fstab entry (essentially) the same as yours (except for mount point, public key path and host of course) I couldn't make it fail.

That should be ssh key path, it's not the public key of the key pair, of course.

Apart from the necessary differences between my entry and yours, your fstab should work and is much the same as the one I used.
Your ssh key does not have a pass phrase, right?

**Comment by sampie** [comment]
>What happens when you run:
>sshfs -o idmap=user,identityfile=/home/sami.ssh/id_rsa sami@lelu.local:/home/sami /mnt/node < /dev/null

The /mnt/node mounts without errors and ls shows properly the files from the mount.

> Your ssh key does not have a pass phrase, right?
No, it does not have a pass phrase. The fstab entry works also fine, if I manually type "mount /mnt/node".

> File system mount helpers live in /usr/sbin not /usr/bin.

My system does not have sshfs file in /usr/sbin. So, should I create a link: /usr/sbin/mount.fuse.sshfs -> /usr/bin/sshfs ?

**Comment by raven-au** [comment]
> > What happens when you run:
> > sshfs -o idmap=user,identityfile=/home/sami.ssh/id_rsa [sami@lelu.local](mailto:sami@lelu.local):/home/sami /mnt/node < /dev/null
> 
> The /mnt/node mounts without errors and ls shows properly the files from the mount.

I have no idea what's going wrong then, (;

I see the autofs file system is compiled as a kernel module in Ubuntu.
Is it loaded, check with "lsmod|grep autofs"

I also see that the loaded module is called autofs4 in Ubuntu although in /proc/filesystems it's registered as autofs, so that should be ok.
But I'm not sure that kernel module will demand load properly (it did for me) although there are some aliases in the module code that are meant to provide for that.

> 
> > Your ssh key does not have a pass phrase, right?
> 
> No, it does not have a pass phrase. The fstab entry works also fine, if I manually type "mount /mnt/node".

Right, mine too.

> 
> > File system mount helpers live in /usr/sbin not /usr/bin.
> 
> My system does not have sshfs file in /usr/sbin. So, should I create a link: /usr/sbin/mount.fuse.sshfs -> /usr/bin/sshfs ?

That's normal, in my Fedora install sshfs is located in /usr/bin as it is in Ubuntu.

But the mount.filesystem_name helpers all live in /usr/sbin and in Fedora I have mount.fuse.sshfs linked to ../bin/sshfs.

I'm guessing that ongoing file system subtype (the .sshfs) handling improvements/implementation means that the fuse.sshfs helper function can be done by the fuse helper and the dependent applications, mount(8), mount.fuse(8) etc. know about this so the extra helper isn't necessary.

**Comment by raven-au** [comment]
Is the host key of lelu.local in the root user's known_hosts host key file?

**Comment by sampie** [comment]
> lsmod|grep autofs

$ lsmod|grep autofs
autofs4                45056  3

> But the mount.filesystem_name helpers all live in /usr/sbin and in Fedora I have 
> mount.fuse.sshfs linked to ../bin/sshfs

I created a link: /usr/sbin/mount.fuse.sshfs -> /usr/bin/sshfs and rebooted the machine. Mount behavior did not change. I still get the error.

> Is the host key of lelu.local in the root user's known_hosts host key file?

/root/.ssh folder is empty.

I received a notification that the bug I have reported to ubuntu launchpad was changed to confirmed, which means other user(s) have been able to reproduce this problem also.

**Comment by raven-au** [comment]
> 
> > Is the host key of lelu.local in the root user's known_hosts host key file?
> 
> /root/.ssh folder is empty.

I was able to make the problem occur by moving root's .ssh/known_hosts out of that directory.

You should be able to have the host key added to the ssh known_hosts file by becoming root and either using ssh to log into the host in question or establishing connection using sshfs from the command line (as root). You should be ask if you want to add it.

**Comment by sampie** [comment]
> I was able to make the problem occur by moving root's .ssh/known_hosts out of that directory.

Interesting. I wonder why that is. Anyway, I became a root and logged into lelu.local with ssh and now the root has known host file with this machine. Then I rebooted machine and ran ls /mnt/node and still got the error.

**Comment by raven-au** [comment]
> > I was able to make the problem occur by moving root's .ssh/known_hosts out of that directory.
> 
> Interesting. I wonder why that is. Anyway, I became a root and logged into lelu.local with ssh and now the root has known host file with this machine. Then I rebooted machine and ran ls /mnt/node and still got the error.

It's because the on-demand mounting is performed as root and that will use that user's known_hosts file to verify the key of the host being connected to, if it isn't a known host then there's no way for the on-demand mounting to know if it's ok to add it, it retries and eventually fails.

**Comment by sampie** [comment]
Makes sense. For some reason, having lelu.local host known by root does not seem to make it work for me.

Did you try with your Ubuntu VM? Can you share your fstab line?

**Comment by sampie** [comment]
I took my original fstab line from the question at askubuntu and then the mount started working. You are right, it looks like root's known_hosts not having the target machine is the root cause of the problem.

**Comment by raven-au** [comment]
> Makes sense. For some reason, having lelu.local host known by root does not seem to make it work for me.
> 
> Did you try with your Ubuntu VM? Can you share your fstab line?

I'm always using the VM since your using Ubuntu.

I have played with two fstab entries both work for me.

One is close to what your using:
aven@centos7:/autofs /mnt/sshfs/centos7 fuse.sshfs noauto,x-systemd.automount,_netdev,users,idmap=user,identityfile=/home/raven/.ssh/nopfid_rsa,allow_other,reconnect 0 0

The other is more like what I would use if I was using sshfs:
raven@centos7:/autofs /mnt/sshfs/centos7 fuse.sshfs noauto,x-systemd.automount,x-systemd.idle-timeout=60,idmap=user,default_permissions,transform_symlinks,identityfile=/home/raven/.ssh/nopfid_rsa,allow_other,uid=1000,gid=1000 0 0

I suspect you need the allow_others option to do the on-demand mounting which means you probably should also use default_permissions and I found that to be able to use the mount as my user without permissions problems I needed the uid= and gid= options.

**Comment by raven-au** [comment]
> I took my original fstab line from the question at askubuntu and then the mount started working. You are right, it looks like root's known_hosts not having the target machine is the root cause of the problem.

As I say, might be that allow_others option.

**Comment by sampie** [comment]
I think how it works (mount as a root) can fulfill what I am aiming to do. However, it is not exactly what I intended to accomplish. I was aiming to have just automatic version of what happens when a specific user types "mount /mnt/node". Now, the manual mount happens as user and automatic happens as root. Is it possible to specify that also this automount is done as specific user? 

My second question is that is it possible to enhance log messages, so that if mount fails for some reason, the logs would give some hint what is wrong?

**Comment by raven-au** [comment]
> I think how it works (mount as a root) can fulfill what I am aiming to do. However, it is not exactly what I intended to accomplish. I was aiming to have just automatic version of what happens when a specific user types "mount /mnt/node". Now, the manual mount happens as user and automatic happens as root. Is it possible to specify that also this automount is done as specific user?

Well, not easily and very likely not securely either.

What happens is when a process walks on an automount point the kernel sends a message to the process handling automounting, the user process then sleeps until the kernel is notified of mount completion which carries a result status. That automounting process, first is distinct, and secondly generally has root privileges so that it can do whatever mounting task is asked of it. There is more to this but I'm not going to go into it now.

The question is, is it worth it, in this case there is the uid= and gid= options and by the look of what your trying to do your going to need to add an entry for each one of these mounts anyway. So automating this by implementing some simple macro substitution (I don't think that exists with systemd automounting) for a few predefined values such as uid and gid isn't going to get you much as you still probably need some more sophisticated substitutions for the mount location string.

> 
> My second question is that is it possible to enhance log messages, so that if mount fails for some reason, the logs would give some hint what is wrong?

The lack of adequate logging is something that this report highlights but the sub-systems involved with this in systemd are several and while they are well defined and versatile they are quite complex. For example, it's difficult to identify exactly where the mounting gets done because, IIUC, a job gets scheduled based on an internally generated automount unit and the internally generated mount unit that's a dependency of the automount unit. It's not obvious on initial reading of the code what that jobs is or what is does and where it enters the mounting code, another fairly general sub-system. That has to be well understood in order to fix the incorrect status return and to improve logging in error cases. Think about it, if the code doesn't think an error has occurred then it will think there's nothing to say!

Unfortunately I have spent way too much time on this already and I have other priorities so investigating it further will need to go onto the todo list.

**Comment by sampie** [comment]
> Unfortunately I have spent way too much time on this already and I have other priorities 
> so investigating it further will need to go onto the todo list.

I would like to thank you for taking time to resolve this issue! I will add the solution also to the askubuntu site, so that users getting “Too many levels of symbolic links" error would more easily find the solution.

**Comment by raven-au** [comment]
> 
> I would like to thank you for taking time to resolve this issue! I will add the solution also to the askubuntu site, so that users getting “Too many levels of symbolic links" error would more easily find the solution.

Thanks, I'm glad I could be of help.

**Comment by raven-au** [comment]
> > I would like to thank you for taking time to resolve this issue! I will add the solution also to the askubuntu site, so that users getting “Too many levels of symbolic links" error would more easily find the solution.
> 
> Thanks, I'm glad I could be of help.

Btw, if there are any improvements done as a result of this they will need to go via the Pull Request mechanism.
So we probably should close this ... up to you though.

**Comment by sampie** [comment]
Of course personally I now know how to make this work, but as a user I recognize the importance of good error messages that guide users to find and fix the root causes. I think these things, like logging and error messages, seem minor, but they have a major impact on the perceived quality of the software as a whole in the long run. So, if systemd has a TODO list, I would be happy if this was added there, but I don't see this being about a discovered bug, but instead a request to improve the current error reporting and logging with a goal to improve the overall quality of systemd.

**Comment by raven-au** [comment]
> Of course personally I now know how to make this work, but as a user I recognize the importance of good error messages that guide users to find and fix the root causes. I think these things, like logging and error messages, seem minor, but they have a major impact on the perceived quality of the software as a whole in the long run. So, if systemd has a TODO list, I would be happy if this was added there, but I don't see this being about a discovered bug, but instead a request to improve the current error reporting and logging with a goal to improve the overall quality of systemd.

Understood and agreed.
But please understand I myself am new to the systemd code base.

I've come to systemd primarily because of my interest in improving mount table handling, hence my interest in the mount and event handling code. But I find myself returning to the systemd code when I really should be getting on with the time critical things that I am responsible for in other areas, I'm a sucker for interesting systems.

I'm also still struggling to work out how to use the Pull Request change submission system, fact is I haven't been very successful as yet.

And I was attracted to this issue because I am very familiar with kernel automounting having worked on it for many years and hoped I could help.

So I'll return to this when I get on top of my current oustanding work, ;)

**Comment by sampie** [comment]
Like in all voluntary projects, all work is highly appreciated, but not required by any means. I wish you happy coding and thank you for your contributions for the open source community.

**Comment by MostHated** [comment]
> > > Is the host key of lelu.local in the root user's known_hosts host key file?
> > 
> > 
> > /root/.ssh folder is empty.
> 
> I was able to make the problem occur by moving root's .ssh/known_hosts out of that directory.
> 
> You should be able to have the host key added to the ssh known_hosts file by becoming root and either using ssh to log into the host in question or establishing connection using sshfs from the command line (as root). You should be ask if you want to add it.

ffs, I sure wish I had came across this sooner. I have spent several hours over the last few days trying to figure out why in the hell I could not get this working. After reading this, I did a 'sudo sshfs' using the identityfile and what not and rebooted and bam, everything worked perfect. Thanks a ton. :partying_face:
